### PR TITLE
Adding sparse matrix transposes

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**📝 Description**
adding matrix transpose

Please mark the PR with appropriate tags: 
feature

**🆕 Changelog**
adding the possibility to generate the transpose of a matrix

### CODE REVIEW

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 1]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
 
 		return pAconverted;
+	}	
 
+    template< class TDataType, class TIndexType >
+	static typename DistributedCsrMatrix<TDataType, TIndexType>::Pointer Transpose(
+			const DistributedCsrMatrix<TDataType, TIndexType>& rA
+			)	
+    {
+		bool move_to_backend=false; //important!
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:146]: Tabs!!
```suggestion
        bool move_to_backend=false; //important!
```

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 2]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
+		auto offdiag_global_index2 = rA.GetOffDiagonalIndex2DataInGlobalNumbering();
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h:147]:
```suggestion
        auto offdiag_global_index2 = rA.GetOffDiagonalIndex2DataInGlobalNumbering();
```
Tabs!

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 3]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
+        auto pAamgcl = ConvertToAmgcl<TDataType,TIndexType>(rA, offdiag_global_index2, move_to_backend);
 
-	}	
+        auto pAamgcl_transpose = transpose(*pAamgcl);
 
+        auto pAt = ConvertToCsrMatrix<TDataType,TIndexType>(*pAamgcl_transpose, rA.GetComm());
+
+		return pAt;
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
        return pAt;
```
tabs!

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 4]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
+	}
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
    }
```
tabs

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 5]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
         // mfem_assemble_colors
 
 		return pAconverted;
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
        return pAconverted;
```
tabs!

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 6]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
+	}
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
    }	
```
more tabs!

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 7]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
         // mfem_assemble_colors
 
 		return pAconverted;
+	}	
 
+    template< class TDataType, class TIndexType >
+	static typename DistributedCsrMatrix<TDataType, TIndexType>::Pointer Transpose(
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
    static typename DistributedCsrMatrix<TDataType, TIndexType>::Pointer Transpose(
```
more tabs

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 8]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
+			const DistributedCsrMatrix<TDataType, TIndexType>& rA
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
        const DistributedCsrMatrix<TDataType, TIndexType>& rA
```
tabs!

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 9]
```diff
@@ -136,10 +136,23 @@ class AmgclDistributedCSRConversionUtilities
...
+			)
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]:
```suggestion
        )	
```
More tabs!

File: kratos/utilities/amgcl_csr_conversion_utilities.h [hunk 10]
```diff
@@ -80,6 +80,19 @@ class AmgclCSRConversionUtilities
         pAconverted->AssignValueData((TDataType*)(rA.val), rA.nnz);
         return pAconverted;
 	}
+
+    //Note that we deliberately return a unique_ptr as it can be moved to a shared_ptr as needed
+    template< class TDataType, class TIndexType >
+	static typename CsrMatrix<TDataType, TIndexType>::Pointer Transpose(
```
Comment (loumalouomega) [inline kratos/utilities/amgcl_csr_conversion_utilities.h]:
```suggestion
    static typename CsrMatrix<TDataType, TIndexType>::Pointer Transpose(
```

File: kratos/utilities/amgcl_csr_conversion_utilities.h [hunk 11]
```diff
@@ -80,6 +80,19 @@ class AmgclCSRConversionUtilities
...
+			CsrMatrix<TDataType, TIndexType>& rA
```
Comment (loumalouomega) [inline kratos/utilities/amgcl_csr_conversion_utilities.h]:
```suggestion
        CsrMatrix<TDataType, TIndexType>& rA
```

File: kratos/utilities/amgcl_csr_conversion_utilities.h [hunk 12]
```diff
@@ -80,6 +80,19 @@ class AmgclCSRConversionUtilities
...
+			)
```
Comment (loumalouomega) [inline kratos/utilities/amgcl_csr_conversion_utilities.h]:
```suggestion
        )	
```

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h [hunk 13]
```diff
@@ -135,11 +135,24 @@ class AmgclDistributedCSRConversionUtilities
...
-		return pAconverted;
+        return pAconverted;
+    }	
 
+    template< class TDataType, class TIndexType >
+    static typename DistributedCsrMatrix<TDataType, TIndexType>::Pointer Transpose(
+        const DistributedCsrMatrix<TDataType, TIndexType>& rA
+        )	
+    {
+        bool move_to_backend=false; //important!
```
Comment (loumalouomega) [inline kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]: Any const here?

File: kratos/utilities/amgcl_csr_conversion_utilities.h [hunk 14]
```diff
@@ -80,6 +80,19 @@ class AmgclCSRConversionUtilities
...
         return pAconverted;
 	}
+
+    //Note that we deliberately return a unique_ptr as it can be moved to a shared_ptr as needed
+    template< class TDataType, class TIndexType >
+    static typename CsrMatrix<TDataType, TIndexType>::Pointer Transpose(
+        CsrMatrix<TDataType, TIndexType>& rA
+        )	
+    {
+        auto pAamgcl = ConvertToAmgcl<TDataType,TIndexType>(rA);
```
Comment (loumalouomega) [inline kratos/utilities/amgcl_csr_conversion_utilities.h]: const?

File: kratos/mpi/tests/test_distributed_sparse_matrices.py [hunk 15]
```diff
@@ -155,6 +155,19 @@ def test_matrix_construction(self):
 
         if(my_rank == 0):
             self.assertEqual(dotprod, np.dot(reference_spmv_res,reference_spmv_res), 1e-14)
+
+        #test transpose and TransposeSpMV
+        b.SetValue(1.0)
+        y.SetValue(0.0)
+        A.TransposeSpMV(b,y) 
+
```
Comment (loumalouomega) [inline kratos/mpi/tests/test_distributed_sparse_matrices.py:163]:
```suggestion
```

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Missing const?

File: kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h (hunk 13 above)
Comment (RiccardoRossi) [reply kratos/mpi/utilities/amgcl_distributed_csr_conversion_utilities.h]: it makes no harm so i'll add it

File: kratos/mpi/tests/test_distributed_sparse_matrices.py (hunk 15 above)
Comment (RiccardoRossi) [reply kratos/mpi/tests/test_distributed_sparse_matrices.py:163]: done

**Review by loumalouomega — APPROVED** [review]

File: kratos/tests/test_sparse_matrices.py [hunk 16]
```diff
@@ -109,6 +109,18 @@ def test_matrix_assembly(self):
...
+        x.SetValue(1.0)
+        y.SetValue(0.0)
+        A.TransposeSpMV(x,y) 
+
+        At = A.Transpose()
+        y2 =  KratosMultiphysics.SystemVector(y.size())
+        y2.SetValue(0.0)  
+        At.SpMV(x,y2)
+        for i in range(y.Size()):
+            self.assertEqual(y[i], y2[i], 1e-14)
```
Comment (philbucher) [inline kratos/tests/test_sparse_matrices.py:122]: there is also `assertVectorAlmostEqual` :)

https://github.com/KratosMultiphysics/Kratos/blob/6d6d5b943ded5c524753fa6807633312665a3e50/kratos/python_scripts/KratosUnittest.py#L76

**Pull request merged**